[JSC] Optimize String.replace with String / String case more in DFG / FTL

Constellation opened the pull request:
#### 965272b76dc3d55f84cb5702ab51ea62cc122e50
<pre>
[JSC] Optimize String.replace with String / String case more in DFG / FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=244750">https://bugs.webkit.org/show_bug.cgi?id=244750</a>

Reviewed by Saam Barati.

While DFG / FTL has optimization for String#replace(RegExp, String), we didn&apos;t have optimization for String#replace(String, String).
This patch adds that optimization since Speedometer2.1 includes String#replace(String, String).

1. We simply add StringReplace with 3 StringUse specialization in DFG. And we invoke very simple operation function for the fast processing.
   To achieve that, we need to have String.prototype.@@replace and Object.prototype.@@replace watchpoints to ensure that these properties
   are not defined.
2. We add String#replace(String, String) strength reduction. So if everything is constant, we will convert it into constant LazyJSString node.
3. To clean up further, we also introduce watchpoints for RegExp primordial properties. We would like to move to the new implementation leveraging
   this for String#replace(RegExp, String) case in the future too to avoid inserting tryGetById in Fixup phase.

Microbenchmark shows 2x improvement. On Intel machine, Speedometer2.1 gets 0.45% faster. And on AppleSilicon machine, 0.26% faster.
In particular Vanilla-ES2015-Babel-Webpack-TodoMVC gets 2% better, Vanilla-ES2015-TodoMVC gets 1.93% better, and VanillaJS-TodoMVC
gets 2.94% better.

                                      ToT                     Patched

    string-replace-string      707.7036+-2.0437     ^    349.5370+-1.3370        ^ definitely 2.0247x faster

* JSTests/microbenchmarks/string-replace-string.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::setupAdaptiveWatchpoint):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::stringSymbolReplaceWatchpointSet):
(JSC::JSGlobalObject::regExpPrimordialPropertiesWatchpointSet):

Canonical link: <a href="https://commits.webkit.org/254156@main">https://commits.webkit.org/254156@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e213c460af0b46f33b5a4dea0e82f1fcf14f0555

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88224 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/69/builds/32552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19005 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97420 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152886 "Built successfully and passed tests") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/30950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26754 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80390 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92087 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93833 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/30950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74987 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24750 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/30950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79816 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79973 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73727 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14766 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26171 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2915 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/31795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37677 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76571 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33932 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16994 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

saambarati commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -13311,6 +13311,48 @@ void SpeculativeJIT::compileStringReplace(Node* node)
+        && node->child1().useKind() == StringUse
+        && node->child2().useKind() == StringUse
+        && node->child3().useKind() == StringUse) {
+        if (JSString* replace = node->child3()->dynamicCastConstant<JSString*>(); replace && !replace->length()) {
```
nit: I think the more canonical JSC thing to do is to have a rule in strength reduction that transforms this node into another node if the replace string is empty. Not super critical to do, but maybe a touch nicer.

Constellation replied:
That sounds good! We are now planning to add more optimization for constant string searching case.
So, in that next patch, I'll convert it to different node in strength-reduction or abstract-interpreter, and then, we will have a specialized optimization for that node's compilation.

saambarati commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -13311,6 +13311,48 @@ void SpeculativeJIT::compileStringReplace(Node* node)
+            m_jit.exceptionCheck();
+            cellResult(result.gpr(), node);
+            if (sample)
+                m_jit.decrementSuperSamplerCount();
```
nit: We do this like 4 or 5 times in this function. If it were me, in the branch at the top of this function where we increment it, I'd make a function that runs on scope exit that emits the decrementing code.

Constellation replied:
That sounds good. We have makeScopeExit for this.

saambarati commented on Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
```diff
@@ -912,16 +916,35 @@ class StrengthReductionPhase : public Phase {
+                size_t matchStart = string.find(searchString);
+                if (matchStart == notFound) {
+                    m_changed = true;
+                    // Preserve any checks we have.
```
nit: this comment isn't adding much, since that's how the code reads

Constellation replied:
OK, dropped.

saambarati reviewed (approved):
r=me

saambarati commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3051,36 +3051,7 @@ bool ByteCodeParser::handleIntrinsicCall(Node* callee, Operand result, Intrinsic
-
-            // Check that searchRegExp[Symbol.match] is still the primordial RegExp.prototype[Symbol.replace]
-            if (!isRegExpPropertySame(globalObject->regExpProtoSymbolReplaceFunction(), m_vm->propertyNames->replaceSymbol.impl()))
+            if (!globalObject->stringSymbolReplaceWatchpointSet().isStillValid() || !globalObject->regExpPrimordialPropertiesWatchpointSet().isStillValid())
```
We never watch `regExpPrimordialPropertiesWatchpointSet`, is that right?

saambarati replied:
Let's add a test, too :-)

Constellation replied:
Sounds good, added.

Constellation replied:
While I added tests to ensure that, it is not bad to just watch it too. So I've added watching code for this watchpoint when we fixup to RegExp path.

Constellation committed:
[JSC] Optimize String.replace with String / String case more in DFG / FTL
https://bugs.webkit.org/show_bug.cgi?id=244750

Reviewed by Saam Barati.

While DFG / FTL has optimization for String#replace(RegExp, String), we didn't have optimization for String#replace(String, String).
This patch adds that optimization since Speedometer2.1 includes String#replace(String, String).

1. We simply add StringReplace with 3 StringUse specialization in DFG. And we invoke very simple operation function for the fast processing.
   To achieve that, we need to have String.prototype.@@replace and Object.prototype.@@replace watchpoints to ensure that these properties
   are not defined.
2. We add String#replace(String, String) strength reduction. So if everything is constant, we will convert it into constant LazyJSString node.
3. To clean up further, we also introduce watchpoints for RegExp primordial properties. We would like to move to the new implementation leveraging
   this for String#replace(RegExp, String) case in the future too to avoid inserting tryGetById in Fixup phase.

Microbenchmark shows 2x improvement. On Intel machine, Speedometer2.1 gets 0.45% faster. And on AppleSilicon machine, 0.26% faster.
In particular Vanilla-ES2015-Babel-Webpack-TodoMVC gets 2% better, Vanilla-ES2015-TodoMVC gets 1.93% better, and VanillaJS-TodoMVC
gets 2.94% better.

                                      ToT                     Patched

    string-replace-string      707.7036+-2.0437     ^    349.5370+-1.3370        ^ definitely 2.0247x faster

* JSTests/microbenchmarks/string-replace-string.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::setupAdaptiveWatchpoint):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::stringSymbolReplaceWatchpointSet):
(JSC::JSGlobalObject::regExpPrimordialPropertiesWatchpointSet):

Canonical link: https://commits.webkit.org/254156@main
